### PR TITLE
Optimize metadata generation through @metadata_reactor.provides()

### PR DESCRIPTION
**Pull request by trehn**

### CODE REVIEW

File: bundlewrap/metagen.py [hunk 1]
```diff
@@ -19,6 +19,180 @@
...
+def reactors_for_paths(available_reactors, required_paths):
+    """
+    Returns only those available_reactors that might affect the
+    required_paths.
+    """
+    for name, reactor in available_reactors:
+        provides = getattr(reactor, '_provides', tuple())
+        if provides:
+            for path in provides:
+                if required_paths.covers(path):
```
Comment (vain) [inline bundlewrap/metagen.py:71]: Shouldn’t this be the other way around? The path from the decorator should cover one of the required paths?

**Comment by vain** [comment]
Still seeing a couple of crashes in our big repo. I’m in the process of debugging.

File: bundlewrap/metagen.py (hunk 1 above)
Comment (trehn) [reply bundlewrap/metagen.py:71]: It's actually both ways:

```
metadata.get('foo') -> required_paths == {'foo'}
@metadata_reactor.provides('foo/bar') -> foo covers foo/bar, so this reactor is relevant
```

```
metadata.get('foo/bar') -> required_paths == {'foo/bar'}
@metadata_reactor.provides('foo') -> reactor might provide foo/bar, so it's relevant
```

But you're right, the latter case is not handled properly.

Comment (trehn) [reply bundlewrap/metagen.py]: pushed a quick fix

Comment (vain) [reply bundlewrap/metagen.py]: Aha! Crashes gone. Metadata hash remained the same. 🎉

**Pull request merged**